net: Add new permission `forceinbound` to evict a random unprotected connection if all slots are otherwise full

pinheadmz opened the pull request:
Closes https://github.com/bitcoin/bitcoin/issues/8798

Use case: I run a full node that accepts inbound connections and have a `whitebind` setting so my personal light client can always connect, even when `maxconnections` (and particularly all inbound slots) is already full.

Currently when connections are full, if we receive an inbound peer request, we look for a current connection to evict so the new peer can have a slot. To find an evict-able peer we go through all our peers and "protect" multiple categories of peers, then we evict the "worst" peer that is left unprotected. If there are no peers left to evict, the inbound connection is denied.

With this PR, if the inbound connection has `forceinbound` permission we start the eviction process by first protecting all `noban` and outbound connections, then selecting one of the remaining peers (if any) at random. Then we loop through all our current connections, removing protected peers from the evict-able list. If we end up protecting all our remaining connections, the randomly chosen peer is evicted.

`forceinbound` implies `noban` permission.

All outbound and `noban` connections remain protected from eviction.

pinheadmz committed:
net: nodes with ForceInbound permission force eviction

pinheadmz committed:
test: cover ForceInbound permission success even when connections are full

stickies-v commented on src/node/eviction.cpp:
```diff
@@ -172,39 +183,56 @@ void ProtectEvictionCandidatesByRatio(std::vector<NodeEvictionCandidate>& evicti
+    if (last.has_value()) last_out = last.value();
+
+    if (vEvictionCandidates.empty()) {
+        if (force) {
+            return last_out.id;
+        } else {
+            return std::nullopt;
+        }
```
nit
```suggestion
        return force ? std::make_optional(last_out.id) : std::nullopt;
```

stickies-v replied:
Not really familiar with this stuff, but that seems to be an MSVC bug? https://stackoverflow.com/questions/55152552/visual-studio-not-performing-rvo-when-ternary-operator-is-used-and-move-copy-cto

We do ternary operators in plenty of places, don't think we need to workaround this?

LarryRuane replied:
Does RVO even apply here? I thought it applies to complex structures, a large vector being the canonical example, where we don't want to copy every element. But here we're just returning a `NodeId` which is a `int64_t`, wrapped by `std::optional`. I don't think there's any savings by moving instead of copying one of these.

stickies-v reviewed (commented):
Concept ACK, makes sense to prioritize whitelisted peers. Will review more in-depth soon.

stickies-v commented on src/node/eviction.cpp:
```diff
@@ -172,39 +183,56 @@ void ProtectEvictionCandidatesByRatio(std::vector<NodeEvictionCandidate>& evicti
     // An attacker cannot predict which netgroups will be protected
-    EraseLastKElements(vEvictionCandidates, CompareNetGroupKeyed, 4);
+    auto last = EraseLastKElements(vEvictionCandidates, CompareNetGroupKeyed, 4);
+    if (last.has_value()) last_out = last.value();
```
An alternative approach to the duplication of these lines is to just store all the return values in a vector, like:

```c++
std::vector<std::optional<NodeEvictionCandidate>> removed;
...
removed.emplace_back(EraseLastKElements(vEvictionCandidates, CompareNetGroupKeyed, 4));
```

and then just return the last one at the end?

stickies-v replied:
An alternative approach to the duplication of these lines is to just store all the return values in a vector, like:

```c++
std::vector<std::optional<NodeEvictionCandidate>> removed;
...
removed.emplace_back(EraseLastKElements(vEvictionCandidates, CompareNetGroupKeyed, 4));
```

and then just return the last one at the end?

stickies-v commented on src/node/eviction.cpp:
```diff
@@ -73,15 +73,22 @@ struct CompareNodeNetworkTime {
-    std::vector<T>& elements, Comparator comparator, size_t k,
+//! Sort an array of NodeEvictionCandidates by the specified comparator, then erase the last K elements where predicate is true.
+//! Return the last element removed if any.
+static std::optional<NodeEvictionCandidate> EraseLastKElements(
```
Returning just the last removed element feels like a very bespoke and perhaps unintuitive interface to me. I know we don't currently need it, but perhaps it's worth generalizing this a bit more and returning all the removed elements instead of just the last one? For the callsite, it's trivial to keep just the last one? Just thinking out loud, perhaps this is difficult to do without an unacceptable overhead in terms of allocations etc.

stickies-v commented on src/net.cpp:
```diff
@@ -925,7 +925,7 @@ bool CConnman::AttemptToEvictConnection()
         }
     }
-    const std::optional<NodeId> node_id_to_evict = SelectNodeToEvict(std::move(vEvictionCandidates));
+    const std::optional<NodeId> node_id_to_evict = SelectNodeToEvict(std::move(vEvictionCandidates), force);
```
nit (and in quite a few other places):
```suggestion
    const std::optional<NodeId> node_id_to_evict{SelectNodeToEvict(std::move(vEvictionCandidates), force)};
```

stickies-v commented on src/node/eviction.cpp:
```diff
@@ -172,39 +183,56 @@ void ProtectEvictionCandidatesByRatio(std::vector<NodeEvictionCandidate>& evicti
+
+    if (vEvictionCandidates.empty()) {
+        if (force) {
+            return last_out.id;
```
All the `last_out` assignments are behind an `if` condition, I think this theoretically can be a zero-initialized `NodeEvictionCandidate`?

stickies-v commented on src/net.h:
```diff
@@ -982,7 +982,7 @@ class CConnman
     bool AlreadyConnectedToAddress(const CAddress& addr);
 
-    bool AttemptToEvictConnection();
+    bool AttemptToEvictConnection(bool force);
```
I think it would be nice to document the effect of `force` both here and for `SelectNodeToEvict`

mzumsande reviewed (commented):
Concept ACK

As I suggested in the review club, an alternative, more simple approach would be to just pick a random peer after removing NoBan and outbound connections when in force-mode. Then, if at the end of the usual eviction algorithm, we don't have a evict-able peer, we would evict the random one instead.
This would make the code simpler (no need to change `EraseLastKElements` or keep track of `last`), and I don't really see a major downside:
* The `EraseLastKElements` eviction criteria don't really seem to be sorted by priority anyway right now (because order doesn't matter so far), so evicting a peer that would've been protected by later calls doesn't seem better than evicting one that would've been protected by earlier ones.
* Whitebind No-Ban peers are more trusted to begin with, but if they maliciously wanted to take over all inbound slots by repeatedly connecting to us on the whitebind address, they can do that anyway, whether the eviction is random or whether the current approach of the PR is used. So it's not clear to me what the extra benefit of the current approach over random eviction is.

pinheadmz committed:
eviction: track one random unprotected node to evict if forced

Accomplished by adding a bool argument `force` to SelectNodeToEvict()

pinheadmz commented:
@stickies-v @mzumsande thanks for youre review and feedback, I refactored the PR so we evict a random peer when forced if all other peers are protected. It is MUCH simpler ;-)

stickies-v commented on src/node/eviction.cpp:
```diff
@@ -175,14 +176,22 @@ void ProtectEvictionCandidatesByRatio(std::vector<NodeEvictionCandidate>& evicti
 
+    // Hang on to one random node to evict if forced
+    std::optional<NodeId> force_evict;
+    if (vEvictionCandidates.size() > 0) {
```
This block can be skipped if `!force`

```suggestion
    if (vEvictionCandidates.size() > 0 && force) {
```

stickies-v commented on src/node/eviction.cpp:
```diff
@@ -175,14 +176,22 @@ void ProtectEvictionCandidatesByRatio(std::vector<NodeEvictionCandidate>& evicti
+    // Hang on to one random node to evict if forced
+    std::optional<NodeId> force_evict;
+    if (vEvictionCandidates.size() > 0) {
+        FastRandomContext rng;
+        size_t randpos = rng.randrange(vEvictionCandidates.size());
```
nit:
```suggestion
        size_t randpos{FastRandomContext().randrange(vEvictionCandidates.size())};
```

stickies-v commented on src/node/eviction.cpp:
```diff
@@ -204,7 +213,8 @@ void ProtectEvictionCandidatesByRatio(std::vector<NodeEvictionCandidate>& evicti
     ProtectEvictionCandidatesByRatio(vEvictionCandidates);
 
-    if (vEvictionCandidates.empty()) return std::nullopt;
+    if (vEvictionCandidates.empty())
+        return force ? force_evict : std::nullopt;
```
Since `force_evict` is already an `optional` defaulting to `nullopt`, this can be simplified in conjunction with my other suggestion.
```suggestion
    if (vEvictionCandidates.empty()) return force_evict;
```

stickies-v commented on src/node/eviction.h:
```diff
@@ -39,7 +39,7 @@ struct NodeEvictionCandidate {
  * remain, the selection logic chooses a peer to evict.
  */
-[[nodiscard]] std::optional<NodeId> SelectNodeToEvict(std::vector<NodeEvictionCandidate>&& vEvictionCandidates);
+[[nodiscard]] std::optional<NodeId> SelectNodeToEvict(std::vector<NodeEvictionCandidate>&& vEvictionCandidates, bool force = false);
```
Documenting the `force` parameter (or linking to `AttemptToEvictConnection`) would be helpful I think.

stickies-v commented on src/net.h:
```diff
@@ -982,7 +982,13 @@ class CConnman
+    /**
+     * Attempt to find a connected node to disconnect.
+     * Used to make room for new inbound connections, returns true if successful.
+     * @param[in] force     Try to evict a random node if all connections are otherwise protected.
```
suggested rephrasing (although I'm not sure about the "NoBan" qualifier, maybe we have a better established term?)
```suggestion
     * @param[in] force     If all connections are otherwise protected, still evict a random inbound NoBan node if it exists
```

stickies-v reviewed (commented):
Light approach ACK c826187b5070ce89edcde0536183714a1c2e3207

I think this random selection approach suggested by mzumsande is a more elegant approach. "Light" in approach ACK because I need to build more confidence that this doesn't introduce new attack vectors.

mzumsande commented on src/net.h:
```diff
@@ -982,7 +982,13 @@ class CConnman
 
-    bool AttemptToEvictConnection();
+    /**
+     * Attempt to find a connected node to disconnect.
```
Maybe just copy over the doc from net.cpp? That's basically what I did in a WIP [branch](https://github.com/mzumsande/bitcoin/commit/1aaa795e3f43caf8aa06331e61b14b777a04a6d9).

stickies-v replied:
I'm not sure that's a strict improvement. Documenting the interface (i.e. pretty much just the first line of what's in `net.cpp`) seems more appropriate to be in the header indeed, but all the rest sounds more like implementation details that may best be kept close to the source code? 

Also, my main issue is with (having just) the word "find", which to me sounds like there are no side effects.

pinheadmz commented:
Thanks @stickies-v nits addressed! 🙏

stickies-v commented on test/functional/p2p_eviction.py:
```diff
@@ -122,6 +122,36 @@ def run_test(self):
 
+        self.log.info("Test that whitebind inbounds get extra eviction power")
+        # Only allow 1 inbound connection, but set whitebind
+        self.restart_node(0, extra_args=['-maxconnections=12', '-whitebind=0.0.0.0:30201'])
```
As you [mentioned](https://bitcoincore.reviews/27600#l-165) in the review club, whitelisting just based on ports enables an attacker that discovered which port(s) are whitelisted to take over all (non-whitelisted) inbound connections slots.

Therefore, I think we have to be careful not to accidentally lead people to this footgun. Binding to `127.0.0.1` seems much more prudent to not set a bad example.

```suggestion
        self.restart_node(0, extra_args=['-maxconnections=12', '-whitebind=127.0.0.1:30201'])
```

pinheadmz replied:
updated test and also included warning in release note

stickies-v reviewed (commented):
Code review ACK 96b513f605fb2df441b66da583056b1c8acd4dbc, but I think the first commit should be removed from this PR, I think it's an accident?

I think the code is good and suits the intent of the PR well. My only concern is about the potential footgun introduced, even if it is relatively mild. Anyone running bitcoind with `whitebind=0.0.0.0:<port>` will now be vulnerable to having all their inbound non-whitelisted slots taken over by an attacker that has figured out what the whitelisted port is. It doesn't seem like a huge concern, given that:
- it only affects inbound connections, which we already assume to be more susceptible to these kinds of attacks, and
- it does require the user to manually whitebind to `0.0.0.0`

but I just wanted to flag it here anyway in case it is actually more serious than I see it to be.

Also, this seems like behaviour change that would benefit from a mention in the release notes?

pinheadmz commented on src/node/eviction.cpp:
```diff
@@ -175,14 +176,21 @@ void ProtectEvictionCandidatesByRatio(std::vector<NodeEvictionCandidate>& evicti
     ProtectOutboundConnections(vEvictionCandidates);
 
+    // Hang on to one random node to evict if forced
+    std::optional<NodeId> force_evict;
+    if (vEvictionCandidates.size() > 0 && force) {
```
Good point, although "if empty, return now" logic would make sense after each step in this function (why bother calling `EraseLastKElements()` with an empty array?) even without this PR. One thing we could do is add that logic to the beginning of `EraseLastKElements()` itself, but I dunno how much time is really wasted in there (`sort`, `min`, `erase` ... ?)

pinheadmz replied:
good point, i'll add it. one line i think in this case will improve readability

pinheadmz committed:
doc: add release note for #27600

stickies-v commented on doc/release-notes-27600.md:
```diff
@@ -0,0 +1,10 @@
+  permissions will now be more likely to connect even if `maxconnections` is
+  reached and the node's inbound slots are full. Because this is achieved by
+  more aggressively finding a current connection to evict, users should take
+  care only to set these permissions on local ports that attackers can not
```
I think "on local ports" is unnecessarily restrictive/scary. I think it's still completely fine to whitelist remote addresses, you mostly just want to avoid ranges, right?

pinheadmz replied:
@mzumsande what's your opinion on this? Is this PR a big enough change to warrant this kind of warning?

mzumsande replied:
Not really sure, I've never looked into these options much and don't know about best practices - I think that if you use `-whitebind` with non-local addresses, you'd at least have to make sure that that address is not self-advertised. I guess that this applies more to `-whitebind` than `-whitelist`, so that the preferred approach in case of a non-local connection would be to use `-whitelist`?
fyi @vasild, do you have an opinion on this?

pinheadmz replied:
That's a good point, `whitelist` is harder to attack than `whitebind` because the attacker would have to spoof their origin IP repeatedly to fill up your inbounds. If a user `whitebind`'s a port and an attacker figures out that port numnber, they can trivially evict all your other inbounds

vasild replied:
I agree with what you say above - `whitebind` on publicly accessible address:port with this new permission sounds bad. Similar with `whitelist` - if a range is used.

This PR currently expands the semantic of the `noban` permission. This will affect existent setups that already use it. Would it make sense to introduce a new permisson, separate from `noban`? I mean - now if somebody is running `-whitebind=noban@publicaddr:port` then a bad actor could cause harm on `master`, but even more harm with this PR.

pinheadmz replied:
@LarryRuane you are correct but what we are changing (in the current state of this branch) is that `noban` nodes can now force disconnection of other peers. Since many users may already have `noban` set, maybe even for a large range of IPs, this branch would introduce a new vulnerability they may not be aware of. Because of that I think it does make more sense to specify a new permission so users can narrow the attack surface

LarryRuane commented:
Please consider editing the description slightly, if my understanding is correct.

>With this PR, if the inbound connection is on our whitelist we start the eviction process by selecting a random unprotected peer.

At this point in the process, where we choose a random peer, we're considering _all_ (inbound, not noban) peers, is that correct? They are all (other than outbound or noban) "unprotected" because we haven't protected any of them yet. So maybe this should say, "... selecting a random peer" or "... selecting a random inbound peer" -- because we may randomly choose a peer that _ends up_ being protected, or not protected. That would make this PR easier for me to understand.

I do think the updated algorithm is very elegant (great suggestion, @mzumsande). It chooses a random peer, which _may_ end up being erased from the eviction candidate list, _or may not_, but we don't care either way; we evict it if we need to, all the better if it actually did not get protected (but, again, we don't care).

By the way, I'm going to highlight this PR in the [Optech newsletter](https://bitcoinops.org/en/newsletters/) next week, that's why I'm particularly interested in making sure I understand it.

LarryRuane commented on src/node/eviction.h:
```diff
@@ -38,8 +38,10 @@ struct NodeEvictionCandidate {
+ *                      are available. (see CConman::AttemptToEvictConnection())
  */
-[[nodiscard]] std::optional<NodeId> SelectNodeToEvict(std::vector<NodeEvictionCandidate>&& vEvictionCandidates);
+[[nodiscard]] std::optional<NodeId> SelectNodeToEvict(std::vector<NodeEvictionCandidate>&& vEvictionCandidates, bool force = false);
```
I'm not a fan of default arguments; consider making this not have a default. It makes sense if there are many existing calls to a function and you don't want to touch them all, but there's only one call to this function in production code. You would have to touch about 6 call sites in test code, but that's not too bad. The reason I'm not a fan of default arguments is that they hide something that may be helpful to see when just looking at the call site. (You might think, "Oh, it's possible that eviction can be 'forced', what does that mean?) But this is an optional suggestion, fine if you leave it as-is, just something to consider.

pinheadmz replied:
ok I like this, will do

naumenkogs replied:
This still remains default, no?

LarryRuane reviewed (commented):
utACK fa78fc543cd46ee1c7181ecf6b696c2892b9f8d3
except for possible attack vectors (for example, `NoBanForce`) that I'm not really qualified to comment on. I may try to contribute to that discussion after I've learned more. Nice PR!

LarryRuane commented:
Even though not needed for this PR, you may want to consider adding a commit to remove the unnecessary templating. It's perfectly reasonable not to make this change in this PR, but I thought I'd document it just in case; I do think it makes the code easier to understand.
```
--- a/src/node/eviction.cpp
+++ b/src/node/eviction.cpp
@@ -74,9 +74,10 @@ struct CompareNodeNetworkTime {
 };
 
 //! Sort an array by the specified comparator, then erase the last K elements where predicate is true.
-template <typename T, typename Comparator>
 static void EraseLastKElements(
-    std::vector<T>& elements, Comparator comparator, size_t k,
+    std::vector<NodeEvictionCandidate>& elements,
+    std::function<bool(const NodeEvictionCandidate&, const NodeEvictionCandidate&)> comparator,
+    size_t k,
     std::function<bool(const NodeEvictionCandidate&)> predicate = [](const NodeEvictionCandidate& n) { return true; })
 {
     std::sort(elements.begin(), elements.end(), comparator);
```

pinheadmz commented:
@LarryRuane thanks for reviewing. I actually had removed the templating in the [very first version](https://github.com/bitcoin/bitcoin/compare/e71d495ffbda3bc072bbaecd7580809d5087f9e6..c826187b5070ce89edcde0536183714a1c2e3207#diff-2404112ebf57bee5f9a16f1a6e1ecfc27a981d37a1c0ff202b4cd9bdfa3e48ccL77) of this branch but after changing the approach I just left it alone. The function is written like a generic utility function, even though we don't currently use it for any other vectors. So I think the function could go as written into a util.cpp or something, in a future PR ?

pinheadmz commented:
@vasild @mzumsande Added a commit that restricts the aggressive eviction to new net permission `ForceInbound` which leaves `NoBan` unchanged.

LarryRuane commented on src/test/fuzz/node_eviction.cpp:
```diff
@@ -40,7 +40,7 @@ FUZZ_TARGET(node_eviction)
     // indeterminate state after the SelectNodeToEvict(&&) call.
     const std::vector<NodeEvictionCandidate> eviction_candidates_copy = eviction_candidates;
-    const std::optional<NodeId> node_to_evict = SelectNodeToEvict(std::move(eviction_candidates));
+    const std::optional<NodeId> node_to_evict = SelectNodeToEvict(std::move(eviction_candidates), /*force=*/false);
```
(I didn't test that this is okay)
```suggestion
    const std::optional<NodeId> node_to_evict = SelectNodeToEvict(std::move(eviction_candidates), /*force=*/fuzzed_data_provider.ConsumeBool());
```

pinheadmz replied:
yeah thanks good diea

LarryRuane reviewed (commented):
utACK 2a6a981e4d892c3d7196ddce54471350950affc7
These latest changes look good! I'll test later in the week.

stickies-v commented on doc/release-notes-27600.md:
```diff
@@ -0,0 +1,10 @@
+P2P
+---
+
+- A new net permission `ForceInbound` (set with `-whitelist=forceinbound@...`
+  or `-whitebind=forceinbound@...`) is introduced that extends `NoBan` permissions.
+  Inbound connections from hosts protected by `ForceInbound` will now be more
```
nit: the capitalized spelling is an internal implementation detail, would stick to `forceinbound` and `noban` here
```suggestion
- A new net permission `forceinbound` (set with `-whitelist=forceinbound@...`
  or `-whitebind=forceinbound@...`) is introduced that extends `noban` permissions.
  Inbound connections from hosts protected by `forceinbound` will now be more
```

pinheadmz replied:
done, thanks

stickies-v reviewed (commented):
Code review ACK 2a6a981e4d892c3d7196ddce54471350950affc7 (modulo my commit reordering comment at the end)

By adding the `forceinbound` (extending `noban`) permission we prevent existing setups from being exposed to a wider attack surface when they have whitelisted ranges of IP addresses. I find it difficult to judge whether the increased complexity (not so much extra code, but more permissions means potentially more interactions to test for etc; as well as the user having to make more choices) is really necessary, but I can see this being a pretty common use case - and this is an elegant solution.

I think the commits should be reordered though, as in https://github.com/stickies-v/bitcoin/commits/pr/27600: I would first introduce "net: add new permission `ForceInbound`" so we don't need to immediately update the logic, test and docs to go from `NoBan` -> `ForceInbound`. Doesn't make things more complex from a re-review pov, imo (range-diff works well for me), and I think it's bad practice to introduce "vulnerabilities" into master, e.g. in case of revert. The rebase is straightforward too.

pinheadmz committed:
net: add new permission ForceInbound

Only inbound nodes with this permission set will call
`SelectNodeToEvict()` with force=true, so when connections are full
there is an increased liklihood of opening a slot for the new inbound.
Extends NoBan permission.

pinheadmz commented:
@stickies-v and @LarryRuane thanks for your review. I re-ordered the commits so the history makes more sense now. Also edited the title and description of the PR since we are adding an optional feature now, not changing any existing behavior. 

`git diff 2a6a981e4d c8ce23745a` reflects the addressed comments, range-diff is screwed up because of commit reordering

stickies-v reviewed (commented):
Code Review re-ACK c8ce23745a

No changes but the review comments from LarryRuane and myself.

luke-jr commented:
Would be better to score the peers rather than choose one at random.

pinheadmz commented:
> Would be better to score the peers rather than choose one at random.

I had logic for this initially but preferred [this suggestion](https://github.com/bitcoin/bitcoin/pull/27600#pullrequestreview-1431340479) by @mzumsande

mzumsande reviewed (commented):
Do I understand it correctly that using the new `forceinbound` permission instead of `noban` would be useful in the following two cases:
1.) If the user lowered `-maxconnections` to a value so low that the protection logic (that protects a fixed number between 20 and 28 inbound peers from eviction) could protect all inbound peers, but we still want to make sure to accept an inbound connection from a protected peer.
2) If we have a slightly higher `-maxconnections` but want to make sure a larger number of protected peers can connect  to us at the same time.

So that a user running with the default `-maxconnections` of 125 would never have a reason to use `forceinbound` over `noban` unless they anticipate to have a large number of protected peers (>80 I guess?) simultaneously?

pinheadmz commented:
@mzumsande This summary looks correct to me. Point (1) is covered by the functional test. You are also right that by default most nodes will not need the flag since there will be enough open slots or evictable peers.

I think the use case highlighted in the issue (as well as my own use case) is more about "personal" nodes or limited-resource nodes, where max connections are more limited but we still always want our light clients to be able to connect.

naumenkogs commented:
Consider when a malicious Internet provider or something can request the `forceinbound` IP address. 

Does this PR open any fundamentally new risk in this case? I'm thinking of disrupting the network by causing many disconnections if the `forceinbound` practice is broad.

This probably depends on whether it's possible to take multiple slots from the same `forceinbound` IP (or, rather, initiate multiple evictions), which I'm not 100% sure about.

If my suspicions are correct, this might better wait for authenticated connections...

pinheadmz commented:
Thanks so much for taking a look Gleb and for your insight, yeah I have the same concerns. That attack vector is not super trivial, the attacker needs to both learn and then spoof the protected IP range. The release notes could recommend only using ranges for local subnets, or we could enforce that in the code. We could also change it so`forceinbound` does not evict existing peers but *may* add connections beyond the maximum, or set another internal limit just for `forceinbound`

naumenkogs commented:
>the attacker needs to both learn and then spoof the protected IP range

I think a malicious ISP gets that trivially (at least in the pre-BIP324 world) by simply monitoring the traffic.

In theory, specifying "forceinbound=iprange;n_extra_slots" instead of eviction is probably a good idea, but I'm not sure how ugly the code will look like. I remember max_connections stuff is already sophisticated :(

pinheadmz commented:
I guess allowing more connections beyond the max is a DoS vector, which could potentially crash a node and break all existing connections anyway... hm/

naumenkogs commented:
@pinheadmz In the example I had above `n_extra_slots` controls how many extra connections you allow so it should be safe.

pinheadmz committed:
net: only allow 8 simultaneous forced inbound connections

pinheadmz commented:
@naumenkogs I added one more commit that limits the number of simultaneous forced-inbound connections to 8. I didn't add it as an init option yet but we could if you like that approach. A connection only counts as forced-inbound if it succeeded in evicting another peer, and that is marked by a new bool in `CNodeOptions`

naumenkogs commented:
@pinheadmz yeah this is probably the best attempt at minimizing the problem I mentioned. Whether it's satisfactory is still at question :) I'm looking forward to seeing what others think.

willcl-ark reviewed (approved):
ACK 8585fe3f80

Only optional style nits if you touch again.

I'm pleased to see the final commit limiting the max number of `forceinbound` connections, as otherwise I too would have been a bit apprehensive of a user sharing a whitebind address:port with the world and seeing _all_ their (non-`noban`) inbounds disconnected.

I've experienced the issue this pull should fix myself, usually solving it by having to remotely restart the node and racing other nodes in re-connecting, which is not particularly user-friendly.

Side note: whilst playing with various option combinations I noticed that we cannot set `-listen=0` in combinatation with `-whitebind=...`. On one level this does make sense, but I now wonder whether `-whitebind` (but not `-bind`) could be permitted when `listen=0`, as it could make sense to want to accept only connections of this type?

pinheadmz commented on src/net.h:
```diff
@@ -352,6 +354,8 @@ struct CNodeOptions
     std::unique_ptr<i2p::sam::Session> i2p_sam_session = nullptr;
     bool prefer_evict = false;
+    // True if ForceInbound connection required evicting a peer
+    bool forced_inbound = false;
```
ah thanks, will touch this up and in one other place in the top commit

pinheadmz commented on src/net_permissions.cpp:
```diff
@@ -15,7 +15,8 @@ const std::vector<std::string> NET_PERMISSIONS_DOC{
     "download (allow getheaders during IBD, no disconnect after maxuploadtarget limit)",
-    "addr (responses to GETADDR avoid hitting the cache and contain random records with the most up-to-date info)"
+    "addr (responses to GETADDR avoid hitting the cache and contain random records with the most up-to-date info)",
+    "forceinbound (when connections are full, attempt to evict a random unprotected peer to open a slot; implies noban)"
```
adding "inbound" good idea thanks

pinheadmz commented on test/functional/p2p_eviction.py:
```diff
@@ -122,6 +122,40 @@ def run_test(self):
+        with node.assert_debug_log(["failed to find an eviction candidate - connection dropped (full)"]):
+            node.add_p2p_connection(RejectedPeer(), wait_for_verack=False, dstport=30201)
+
+        self.log.debug("ForceInbound whitebind inbound gets connected, even when full")
```
I'm trying to figure out if
1. Local peers are protected in any special way (I don't think they are?)
2. Inbound onion connections are considered local (They aren't - inbound onions are identifiable because they bind to a different network port `onion_service_target_port`)

And then, I don't think all onion peers are protected per se but their latency is compensated for in the eviction process:

https://github.com/bitcoin/bitcoin/blob/c9f288244b8d183e09a917025922b99e3368ef78/src/node/eviction.cpp#L105-L112

pinheadmz commented:
@willcl-ark thanks for reviewing, addressed your feedback except for the new test but always open to writing more tests so lemme know if you think something is uncovered there.

pinheadmz commented:
Rebased on master again, thanks @willcl-ark for your review. Hoping to get some more feedback from @mzumsande and @naumenkogs to proceed, or abandon

mzumsande reviewed (commented):
> Hoping to get some more feedback from @mzumsande and @naumenkogs to proceed, or abandon

Personally, I'm not sure if the use case (running a node with `-maxconnections < 40` and at the same time wanting certain inbound peers, or wanting a really high number of whitelisted inbounds peers at the same time) is common enough to introduce an extra permission just for that. In more typical cases, `noban` should be sufficient.
But if others think it is I'm not against it.

A related use case I could see (but I'm also not sure if there is demand for that) is to only be reachable by whitelisted "friends" but no one else. That doesn't seem to be possible right now, `-fListen` is either all or none.

naumenkogs commented:
@pinheadmz Honestly I still struggle to understand what are the practical scenarios of using this. E.g., what could lead to not being able to evict a connection?

pinheadmz commented:
@naumenkogs If the user has *also* set a low `maxconnections` value then [`SelectNodeToEvict()`](https://github.com/bitcoin/bitcoin/blob/19d1ba1b4141f744cca291ff2d99d0b8ffcf946d/src/node/eviction.cpp#L186-L197) might return null.

Quick math 4+8+4+8+4 = 28 protected nodes. Plus 8 full outbound and 2 block only = 38. So if a user runs a full node on limited hardware (like my Raspberry Pi at home) and have something like `-maxconnections=30` then even with the existing `whitebind` permissions, they're privileged node may fail to connect.

That being said, it ALSO means that the solution to https://github.com/bitcoin/bitcoin/issues/8798 may be to just ensure `maxconnections` is set > 40. It's been 7 years since that issue was open and I can check with my own (modern) hardware that 40 connections is sane.

naumenkogs commented on src/node/eviction.cpp:
```diff
@@ -175,14 +176,24 @@ void ProtectEvictionCandidatesByRatio(std::vector<NodeEvictionCandidate>& evicti
     ProtectOutboundConnections(vEvictionCandidates);
 
+    // No inbound or unprotected connections left, return early
+    if (vEvictionCandidates.empty()) return std::nullopt;
```
c5e7e6563201965e88e07aa8da8f26e17fc1b4db

"unprotected" here is sloppy. I'd just drop the comment.

pinheadmz replied:
ok

naumenkogs commented on src/node/eviction.cpp:
```diff
@@ -204,7 +215,7 @@ void ProtectEvictionCandidatesByRatio(std::vector<NodeEvictionCandidate>& evicti
     ProtectEvictionCandidatesByRatio(vEvictionCandidates);
 
-    if (vEvictionCandidates.empty()) return std::nullopt;
+    if (vEvictionCandidates.empty()) return force_evict;
```
I'd suggest expanding this comment to explain what exactly happened.

naumenkogs commented on src/node/eviction.h:
```diff
@@ -38,8 +38,10 @@ struct NodeEvictionCandidate {
  * ratios of desirable or disadvantaged peers. If any eviction candidates
  * remain, the selection logic chooses a peer to evict.
+ * @param[in]   force   Attempt to evict a random peer if no candidates
+ *                      are available. (see CConman::AttemptToEvictConnection())
```
c5e7e6563201965e88e07aa8da8f26e17fc1b4db

"among inbound no-noban connections" or something (i hate `no-noban` but not sure what's better)

pinheadmz replied:
done, thanks

naumenkogs commented on src/net.cpp:
```diff
@@ -1776,8 +1776,9 @@ void CConnman::CreateNodeFromAcceptedSocket(std::unique_ptr<Sock>&& sock,
-    int nInbound = 0;
-    int nMaxInbound = nMaxConnections - m_max_outbound;
+    int nInbound{0};
+    int nForced{0};
```
311902f2cf94ee0e11ed34a1373db42dbc218b20

No need to use legacy naming patterns for new variables :)

pinheadmz replied:
hmmm I'm gonna keep it actually because later in this function is a boolean called `forced` so, I think this is legible

naumenkogs commented on src/net.cpp:
```diff
@@ -1865,6 +1877,7 @@ void CConnman::CreateNodeFromAcceptedSocket(std::unique_ptr<Sock>&& sock,
                              CNodeOptions{
                                  .permission_flags = permission_flags,
                                  .prefer_evict = discouraged,
+                                 .forced_inbound = forced,
```
311902f2cf94ee0e11ed34a1373db42dbc218b20

should this be exposed in peer info?

pinheadmz replied:
good idea, added a new commit for this

pinheadmz committed:
net: add forced_inbound to getpeerinfo

pinheadmz commented:
Thanks for the review Gleb, I also rebased on master to fix conflicts

naumenkogs commented on src/net.h:
```diff
@@ -226,6 +226,8 @@ class CNodeStats
     TransportProtocolType m_transport_type;
     /** BIP324 session id string in hex, if any. */
     std::string m_session_id;
+    /** whether this peer forced its connection by evicting another */
```
8c2026848da910fdebff0a9f73e29f1f6ae81e43

So forced actually means two things to you:
1. Before connecting — something that *might* evict another peer
2. After connecting — something that *has* evicted another peer

I think this is confusing and better to use different words. RPC can expose "force_evicted" flag (to cover the latter case and apply to the `Connection`), while "forceInbound" can stay in the network/potential-connection context (the former case).

naumenkogs replied:
Apparently, in the previous commit 75868022a904c1f77871abf962bf9b88a9c5faf6 you assign `forced = true;` even for non-`forceInbound` peers. 

So basically if a node had 8 connections which are non-`forceInbound` but still did eviction, a real `forceInbound` connection won't be able to join (see how you count nForced). Is that intended? Seems like a bug to me.

naumenkogs reviewed (commented):
Not sure if you missed the two comments:
[one](https://github.com/bitcoin/bitcoin/pull/27600/commits/0c0f2a2c3664d5ea81ee31230ddca40863dd76dd#r1387600158)
[two](https://github.com/bitcoin/bitcoin/pull/27600/commits/0c0f2a2c3664d5ea81ee31230ddca40863dd76dd#r1387598983)

sr-gi commented:
I feel like this makes sense conceptually, but I have similar concerns to @stickies-v @mzumsande and @naumenkogs with it. 

If we could approach this without the need to introduce an additional permission I'd be happier, since it seems a big change for a narrow use case with a potential workaround by just accepting more than 38 total connections (plus `N` for the nobans). I also wonder how this plays out with #27114, given this would be an `in` only permission. 

Given this only triggers under really specific conditions, can we not just prioritize our whitelisted peer under those conditions? The addition of new whitelisted peers can take priority over the ones we protect for "diversity criteria" as long as we do not have enough candidates to evict a peer, that is: after protecting the current `noban` and outbound peers, if the number of remaining peers is smaller than the number of peers we will protect for diversity criteria, we just pick one of the remaining at random and remove them. All this limited to a maximum as you are already doing with `MAX_FORCED_INBOUND_CONNECTIONS`. 

For this to trigger, your number of `noban` connections needs to be too high, and/or, your max number of connections is too low. If we are concerned that this could backfire, as @vasild mentioned, we could even gatekeep it under a global flag so it does not affect any of the existing `noban` connections.
